// @okta/okta-auth-js/cjs/idx/remediators/AuthenticatorVerificationData.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

exports.AuthenticatorVerificationData = void 0;

var _reduce = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/reduce"));

var _find = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/find"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

var _errors = require("../../errors");

var _AuthenticatorData = require("./Base/AuthenticatorData");

/* eslint-disable @typescript-eslint/no-non-null-assertion */

/*!
 * Copyright (c) 2015-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * 
 * See the License for the specific language governing permissions and limitations under the License.
 */
class AuthenticatorVerificationData extends _AuthenticatorData.AuthenticatorData {
  constructor(remediation, values = {}) {
    super(remediation, values); // TODO: extend this feature to all authenticators

    this.shouldProceedWithEmailAuthenticator = this.authenticator.methods.length === 1 && this.authenticator.methods[0].type === 'email';
  }

  canRemediate() {
    // auto proceed if there is only one method
    if (this.shouldProceedWithEmailAuthenticator) {
      return true;
    }

    return super.canRemediate();
  }

  mapAuthenticator() {
    var _context2, _value$find, _context3;

    const authenticatorData = this.getAuthenticatorData();
    const authenticatorFromRemediation = this.getAuthenticatorFromRemediation(); // auto proceed with the only methodType option

    if (this.shouldProceedWithEmailAuthenticator) {
      var _authenticatorFromRem, _context;

      return (_authenticatorFromRem = authenticatorFromRemediation.form) === null || _authenticatorFromRem === void 0 ? void 0 : (0, _reduce.default)(_context = _authenticatorFromRem.value).call(_context, (acc, curr) => {
        if (curr.value) {
          acc[curr.name] = curr.value;
        } else if (curr.options) {
          acc[curr.name] = curr.options[0].value;
        } else {
          throw new _errors.AuthSdkError(`Unsupported authenticator data type: ${curr}`);
        }

        return acc;
      }, {});
    } // return based on user selection


    return {
      id: (0, _find.default)(_context2 = authenticatorFromRemediation.form.value).call(_context2, ({
        name
      }) => name === 'id').value,
      enrollmentId: (_value$find = (0, _find.default)(_context3 = authenticatorFromRemediation.form.value).call(_context3, ({
        name
      }) => name === 'enrollmentId')) === null || _value$find === void 0 ? void 0 : _value$find.value,
      methodType: authenticatorData === null || authenticatorData === void 0 ? void 0 : authenticatorData.methodType
    };
  }

  getInputAuthenticator() {
    var _context4;

    const authenticator = this.getAuthenticatorFromRemediation();
    const methodType = (0, _find.default)(_context4 = authenticator.form.value).call(_context4, ({
      name
    }) => name === 'methodType'); // if has methodType in form, let user select the methodType

    if (methodType && methodType.options) {
      return {
        name: 'methodType',
        type: 'string',
        required: true
      };
    } // no methodType, then return form values


    const inputs = [...authenticator.form.value];
    return inputs;
  }

}

exports.AuthenticatorVerificationData = AuthenticatorVerificationData;
(0, _defineProperty2.default)(AuthenticatorVerificationData, "remediationName", 'authenticator-verification-data');
