// @okta/okta-auth-js/cjs/idx/idx-js/interact.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

exports.default = void 0;

var _promise = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/promise"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _entries = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/entries"));

var _client = require("./client");

/*!
 * Copyright (c) 2021-Present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

/* eslint-disable camelcase */
// @ts-nocheck
const parseAndReject = response => response.json().then(err => _promise.default.reject(err));

const interact = async function interact({
  withCredentials,
  clientId,
  baseUrl,
  scopes = ['openid', 'email'],
  redirectUri,
  codeChallenge,
  codeChallengeMethod,
  state,
  activationToken,
  recoveryToken,
  clientSecret
}) {
  var _context;

  const target = `${baseUrl}/v1/interact`;
  const params = {
    client_id: clientId,
    scope: scopes.join(' '),
    redirect_uri: redirectUri,
    code_challenge: codeChallenge,
    code_challenge_method: codeChallengeMethod,
    state
  };

  if (activationToken) {
    params.activation_token = activationToken;
  }

  if (recoveryToken) {
    params.recovery_token = recoveryToken;
  }

  if (clientSecret) {
    params.client_secret = clientSecret;
  }

  const body = (0, _map.default)(_context = (0, _entries.default)(params)).call(_context, ([param, value]) => `${param}=${encodeURIComponent(value)}`).join('&');
  const headers = {
    'content-type': 'application/x-www-form-urlencoded'
  };
  const credentials = withCredentials === false ? 'omit' : 'include';
  return (0, _client.request)(target, {
    credentials,
    headers,
    body
  }).then(response => response.ok ? response.json() : parseAndReject(response)).then(data => data.interaction_handle);
};

var _default = interact;
exports.default = _default;
module.exports = exports.default;
